src/Simplify_Call.cpp

```diff
@@ -125,56 +125,40 @@ Expr Simplify::visit(const Call *op, ExprInfo *bounds) {
-                    ib = -std::max(ib, -i64_max);
+        uint64_t ub = 0;
+        if (const_int(b, &ib)) {
+            // LLVM shl and shr instructions produce poison for negative shifts,
+            // or for shifts >= typesize, so we will follow suit in our simplifier.
+            user_assert(ib >= 0) << "bitshift by a constant negative amount is not legal in Halide.";
+            user_assert(ib < t.bits()) << "bitshift by a constant amount >= the type size is not legal in Halide.";
+            if (ib < t.bits() - 1) {
+                b = make_const(t, ((int64_t) 1LL) << ib);
+                if (shift_left) {
+                    return mutate(Mul::make(a, b), bounds);
                 } else {
-                    // It's a uint > 2^63-1, so we're definitely shifting out-of-range;
-                    // just clip to a large-enough value value and fall thru.
-                    ib = i64_max;
-                }
-            }
-
-            if (ib >= 0) {
-                assert(t.bits() <= 64);
-                int max_shift = t.bits() - 1;
-                if (t.is_int()) {
-                    // One less if there is a sign bit
-                    max_shift -= 1;
+                    return mutate(Div::make(a, b), bounds);
                 }
-                if (ib <= max_shift) {
-                    ib = 1LL << ib;
-                    b = make_const(t, ib);
-
-                    if (shift_left) {
-                        return mutate(Mul::make(a, b), bounds);
-                    } else {
-                        return mutate(Div::make(a, b), bounds);
-                    }
+            } else {
+                // (1 << ib) will overflow into the sign bit, making decomposition into mul or div
+                // probelmatic, so just special-case them here.
+                if (shift_left) {
+                    return mutate(select((a & 1) != 0, make_const(t, ((int64_t) 1LL) << ib), make_zero(t)), bounds);
                 } else {
-                    // We know that all the meaningful bits will be shifted away;
-                    // just replace with a constant 0 or -1 if possible.
-                    if (shift_left) {
-                        return make_zero(t);
-                    } else {
-                        // shift-right-out-of-bounds -> zero for uint.
-                        if (t.is_uint()) {
-                            return make_zero(t);
-                        } else {
-                            // shift-right-out-of-bounds -> zero or -1 for int.
-                            return mutate(select(a < 0, make_const(t, -1), make_zero(t)), bounds);
-                        }
-                    }
+                    return mutate(select(a < 0, make_const(t, -1), make_zero(t)), bounds);
                 }
             }
+        } else if (const_uint(b, &ub)) {
+            user_assert(ub < (uint64_t) t.bits()) << "bitshift by a constant amount >= the type size is not legal in Halide.";
```

Comment:
I predict there are cases that will trip this that aren't due to user code, but rather created by IR manipulation inside the compiler (if not in bounds). We might need to return a quiet failure expression instead (e.g. signed_integer_overflow, though that would probably need to be renamed if used for const uints).

Reply:
That's possible, but I'd be inclined to see how much it happens in practice first; as of right now, public Halide code is clean with this in place. (Checking against the massive Google corpus of Halide code would be instructive.)